c++11/include/opentracing/span.h

```diff
@@ -0,0 +1,138 @@
+#ifndef OPENTRACING_SPAN_H
+#define OPENTRACING_SPAN_H
+
+#include <chrono>
+#include <functional>
+#include <string>
+
+#include <opentracing/util.h>
+#include <opentracing/value.h>
+
+namespace opentracing {
+class Tracer;
+
+// SpanContext represents Span state that must propagate to descendant Spans and
+// across process boundaries (e.g., a <trace_id, span_id, sampled> tuple).
+class SpanContext {
+ public:
+  virtual ~SpanContext() = default;
+
+  // ForeachBaggageItem calls a function for each baggage item in the
+  // context.  If the function returns false, it will not be called
+  // again and ForeachBaggageItem will return.
+  virtual void ForeachBaggageItem(
+      std::function<bool(const std::string& key, const std::string& value)> f)
+      const = 0;
+};
+
+// FinishOptions allows Span.Finish callers to override the finish
+// timestamp.
+struct FinishSpanOptions {
+  SteadyTime finish_steady_timestamp;
+};
+
+// FinishSpanOption instances (zero or more) may be passed to Span.Finish.
+class FinishSpanOption {
+ public:
+  FinishSpanOption(const FinishSpanOption&) = delete;
+
+  virtual ~FinishSpanOption() = default;
+
+  virtual void Apply(FinishSpanOptions& options) const = 0;
+
+ protected:
+  FinishSpanOption() = default;
+};
+
+// Span represents an active, un-finished span in the OpenTracing system.
+//
+// Spans are created by the Tracer interface.
+class Span {
+ public:
+  virtual ~Span() = default;
+
+  // Sets the end timestamp and finalizes Span state.
+  //
+  // With the exception of calls to context() (which are always allowed),
+  // Finish() must be the last call made to any span instance, and to do
+  // otherwise leads to undefined behavior.
+  void Finish(std::initializer_list<option_wrapper<FinishSpanOption>>
+                  option_list = {}) {
+    FinishSpanOptions options;
+    options.finish_steady_timestamp = SteadyClock::now();
+    for (const auto& option : option_list) option.get().Apply(options);
+    FinishWithOptions(options);
+  }
+
+  virtual void FinishWithOptions(
+      const FinishSpanOptions& finish_span_options) = 0;
+
+  // Sets or changes the operation name.
+  //
+  // SetOperationName may be called prior to Finish.
+  virtual void SetOperationName(const std::string& name) = 0;
+
+  // Adds a tag to the span.
+  //
+  // If there is a pre-existing tag set for `key`, it is overwritten.
+  //
+  // Tag values can be numeric types, strings, or bools. The behavior of
+  // other tag value types is undefined at the OpenTracing level. If a
+  // tracing system does not know how to handle a particular value type, it
+  // may ignore the tag, but shall not panic.
+  //
+  // SetTag may be called prior to Finish.
+  virtual void SetTag(const std::string& key, const Value& value) = 0;
```

Comment:
I'm not sure. On the one hand, I think it's useful to allow anything goes here, but probably difficult for people to agree on the semantics / interpretation of nested structure other than scalar values. I'd suggest leaving our options open and getting this around for review.